In [1]:
# Importing Libraries
from selenium.webdriver.common.by import By
from selenium.webdriver.common.action_chains import ActionChains
from selenium.webdriver.common.keys import Keys
from selenium import webdriver
import time
from bs4 import BeautifulSoup

# Initialize WebDriver
driver = webdriver.Chrome()

# Open the URL
url = "https://www.flipkart.com/apple-iphone-15-blue-128-gb/product-reviews/itmbf14ef54f645d?pid=MOBGTAGPAQNVFZZY&lid=LSTMOBGTAGPAQNVFZZYQRLPCQ&marketplace=FLIPKART"
driver.get(url)

time.sleep(2)

Names = []
Reviews = []
Dates = []
Cities = []

for i in range(1, 11):
    time.sleep(3)
    soup = BeautifulSoup(driver.page_source, "html.parser")

    name = soup.find_all("p", {"class": "_2NsDsF AwS1CA"})
    for n in name:
        Names.append(n.text)

    city = soup.find_all("p", {"class": "MztJPv"})
    for c in city:
        Cities.append(c.text)

    date = soup.find_all("p", {"class": "_2NsDsF"})
    for d in date:
        Dates.append(d.text)

    review = soup.find_all("div", {"class": "ZmyHeo"})
    for r in review:
        Reviews.append(r.text)

    time.sleep(3)

    try:
        # Find the next button
        next_button = driver.find_element(By.XPATH, """/html/body/div/div/div[3]/div/div/div[2]/div[13]/div/div/nav/a[6]""") #till page-06

        # Scroll into view
        driver.execute_script("arguments[0].scrollIntoView();", next_button)
        time.sleep(1)

        # Click the next button
        next_button.click()
    except Exception as e:
        print("Could not click next button:", e)
        break  # Stop the loop if pagination fails

    time.sleep(3)

driver.quit()


In [2]:
Dates # all elements in the dates column

['Mousam  Guha Roy',
 'Oct, 2023',
 'Ajin V',
 'Oct, 2023',
 'bijaya mohanty',
 '8 months ago',
 'Prithivi Boruah',
 'Oct, 2023',
 'Nikhil Kumar',
 'Jan, 2024',
 'Flipkart Customer',
 'Jan, 2024',
 'Akshay Meena',
 'Nov, 2023',
 'Sheetla Prasad Maurya',
 'Oct, 2023',
 'Raj Singh',
 'Dec, 2023',
 'Arunji Govindaraju',
 'Feb, 2024',
 'Mukesh Kumar',
 'Oct, 2023',
 'Feroz Basha',
 'Jan, 2024',
 'ANUP SINGH GAUTAM',
 'Oct, 2023',
 'Preenit  Rai',
 'Feb, 2024',
 'PROSIT DAS',
 '9 months ago',
 'Urgain Norzin',
 'Oct, 2023',
 'Surabhi Das',
 'Jan, 2024',
 'Nilesh  Yadav',
 'Sep, 2023',
 'SYAM SUNDER',
 'Jan, 2024',
 'Abhishek Kumar',
 '8 months ago',
 'Mukesh Kumar',
 'Oct, 2023',
 'Feroz Basha',
 'Jan, 2024',
 'ANUP SINGH GAUTAM',
 'Oct, 2023',
 'Preenit  Rai',
 'Feb, 2024',
 'PROSIT DAS',
 '9 months ago',
 'Urgain Norzin',
 'Oct, 2023',
 'Surabhi Das',
 'Jan, 2024',
 'Nilesh  Yadav',
 'Sep, 2023',
 'SYAM SUNDER',
 'Jan, 2024',
 'Abhishek Kumar',
 '8 months ago',
 'Mukesh Kumar',
 'Oct, 202

Fixed Relative Date Conversion: Used datetime to convert terms like ("11 months ago"etc) into actual dates.

In [3]:
import re

# Given list
data =  ['VENKATESH REDDY',
 '11 months ago',
 'Ajin V',
 'Oct, 2023',
 'Mousam  Guha Roy',
 'Oct, 2023',
 'bijaya mohanty',
 '8 months ago',
 'Prithivi Boruah',
 'Oct, 2023',
 'Nikhil Kumar',
 'Jan, 2024',
 'Flipkart Customer',
 'Jan, 2024',
 'Akshay Meena',
 'Nov, 2023',
 'Sheetla Prasad Maurya',
 'Oct, 2023',
 'Raj Singh',
 'Dec, 2023',
 'Mukesh Kumar',
 'Oct, 2023',
 'Feroz Basha',
 'Jan, 2024',
 'ANUP SINGH GAUTAM',
 'Oct, 2023',
 'Preenit  Rai',
 'Feb, 2024',
 'PROSIT DAS',
 '9 months ago',
 'Urgain Norzin',
 'Oct, 2023',
 'Surabhi Das',
 'Jan, 2024',
 'Nilesh  Yadav',
 'Sep, 2023',
 'SYAM SUNDER',
 'Jan, 2024',
 'Abhishek Kumar',
 '8 months ago',
 'Mukesh Kumar',
 'Oct, 2023',
 'Feroz Basha',
 'Jan, 2024',
 'ANUP SINGH GAUTAM',
 'Oct, 2023',
 'Preenit  Rai',
 'Feb, 2024',
 'PROSIT DAS',
 '9 months ago',
 'Urgain Norzin',
 'Oct, 2023',
 'Surabhi Das',
 'Jan, 2024',
 'Nilesh  Yadav',
 'Sep, 2023',
 'SYAM SUNDER',
 'Jan, 2024',
 'Abhishek Kumar',
 '8 months ago',
 'Mukesh Kumar',
 'Oct, 2023',
 'Feroz Basha',
 'Jan, 2024',
 'ANUP SINGH GAUTAM',
 'Oct, 2023',
 'Preenit  Rai',
 'Feb, 2024',
 'PROSIT DAS',
 '9 months ago',
 'Urgain Norzin',
 'Oct, 2023',
 'Surabhi Das',
 'Jan, 2024',
 'Nilesh  Yadav',
 'Sep, 2023',
 'SYAM SUNDER',
 'Jan, 2024',
 'Abhishek Kumar',
 '8 months ago',
 'Mukesh Kumar',
 'Oct, 2023',
 'Feroz Basha',
 'Jan, 2024',
 'ANUP SINGH GAUTAM',
 'Oct, 2023',
 'Preenit  Rai',
 'Feb, 2024',
 'PROSIT DAS',
 '9 months ago',
 'Urgain Norzin',
 'Oct, 2023',
 'Surabhi Das',
 'Jan, 2024',
 'Nilesh  Yadav',
 'Sep, 2023',
 'SYAM SUNDER',
 'Jan, 2024',
 'Abhishek Kumar',
 '8 months ago',
 'Mukesh Kumar',
 'Oct, 2023',
 'Feroz Basha',
 'Jan, 2024',
 'ANUP SINGH GAUTAM',
 'Oct, 2023',
 'Preenit  Rai',
 'Feb, 2024',
 'PROSIT DAS',
 '9 months ago',
 'Urgain Norzin',
 'Oct, 2023',
 'Surabhi Das',
 'Jan, 2024',
 'Nilesh  Yadav',
 'Sep, 2023',
 'SYAM SUNDER',
 'Jan, 2024',
 'Abhishek Kumar',
 '8 months ago',
 'Mukesh Kumar',
 'Oct, 2023',
 'Feroz Basha',
 'Jan, 2024',
 'ANUP SINGH GAUTAM',
 'Oct, 2023',
 'Preenit  Rai',
 'Feb, 2024',
 'PROSIT DAS',
 '9 months ago',
 'Urgain Norzin',
 'Oct, 2023',
 'Surabhi Das',
 'Jan, 2024',
 'Nilesh  Yadav',
 'Sep, 2023',
 'SYAM SUNDER',
 'Jan, 2024',
 'Abhishek Kumar',
 '8 months ago',
 'Mukesh Kumar',
 'Oct, 2023',
 'Feroz Basha',
 'Jan, 2024',
 'ANUP SINGH GAUTAM',
 'Oct, 2023',
 'Preenit  Rai',
 'Feb, 2024',
 'PROSIT DAS',
 '9 months ago',
 'Urgain Norzin',
 'Oct, 2023',
 'Surabhi Das',
 'Jan, 2024',
 'Nilesh  Yadav',
 'Sep, 2023',
 'SYAM SUNDER',
 'Jan, 2024',
 'Abhishek Kumar',
 '8 months ago',
 'Mukesh Kumar',
 'Oct, 2023',
 'Feroz Basha',
 'Jan, 2024',
 'ANUP SINGH GAUTAM',
 'Oct, 2023',
 'Preenit  Rai',
 'Feb, 2024',
 'PROSIT DAS',
 '9 months ago',
 'Urgain Norzin',
 'Oct, 2023',
 'Surabhi Das',
 'Jan, 2024',
 'Nilesh  Yadav',
 'Sep, 2023',
 'SYAM SUNDER',
 'Jan, 2024',
 'Abhishek Kumar',
 '8 months ago',
 'Mukesh Kumar',
 'Oct, 2023',
 'Feroz Basha',
 'Jan, 2024',
 'ANUP SINGH GAUTAM',
 'Oct, 2023',
 'Preenit  Rai',
 'Feb, 2024',
 'PROSIT DAS',
 '9 months ago',
 'Urgain Norzin',
 'Oct, 2023',
 'Surabhi Das',
 'Jan, 2024',
 'Nilesh  Yadav',
 'Sep, 2023',
 'SYAM SUNDER',
 'Jan, 2024',
 'Abhishek Kumar',
 '8 months ago'

]

# Patterns
date_pattern = r'^(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec),\s\d{4}$'  # Month, Year
relative_pattern = r'^\d+\smonths\sago$'  # "8 months ago" format

# Extract both types
dates_only = [item for item in data if re.match(date_pattern, item) or re.match(relative_pattern, item)]

print(dates_only)
print("Total extracted dates:", len(dates_only))


['11 months ago', 'Oct, 2023', 'Oct, 2023', '8 months ago', 'Oct, 2023', 'Jan, 2024', 'Jan, 2024', 'Nov, 2023', 'Oct, 2023', 'Dec, 2023', 'Oct, 2023', 'Jan, 2024', 'Oct, 2023', 'Feb, 2024', '9 months ago', 'Oct, 2023', 'Jan, 2024', 'Sep, 2023', 'Jan, 2024', '8 months ago', 'Oct, 2023', 'Jan, 2024', 'Oct, 2023', 'Feb, 2024', '9 months ago', 'Oct, 2023', 'Jan, 2024', 'Sep, 2023', 'Jan, 2024', '8 months ago', 'Oct, 2023', 'Jan, 2024', 'Oct, 2023', 'Feb, 2024', '9 months ago', 'Oct, 2023', 'Jan, 2024', 'Sep, 2023', 'Jan, 2024', '8 months ago', 'Oct, 2023', 'Jan, 2024', 'Oct, 2023', 'Feb, 2024', '9 months ago', 'Oct, 2023', 'Jan, 2024', 'Sep, 2023', 'Jan, 2024', '8 months ago', 'Oct, 2023', 'Jan, 2024', 'Oct, 2023', 'Feb, 2024', '9 months ago', 'Oct, 2023', 'Jan, 2024', 'Sep, 2023', 'Jan, 2024', '8 months ago', 'Oct, 2023', 'Jan, 2024', 'Oct, 2023', 'Feb, 2024', '9 months ago', 'Oct, 2023', 'Jan, 2024', 'Sep, 2023', 'Jan, 2024', '8 months ago', 'Oct, 2023', 'Jan, 2024', 'Oct, 2023', 'Feb, 2

In [4]:
from datetime import datetime, timedelta

def convert_relative_date(text):
    if re.match(relative_pattern, text):
        months_ago = int(text.split()[0])  # Extract number
        target_date = datetime.today() - timedelta(days=months_ago * 30)  # Approximate month length
        return target_date.strftime("%b, %Y")  # Convert to "Jan, 2024" format
    return text

converted_dates = [convert_relative_date(item) for item in dates_only]
print(converted_dates)


['Mar, 2024', 'Oct, 2023', 'Oct, 2023', 'Jun, 2024', 'Oct, 2023', 'Jan, 2024', 'Jan, 2024', 'Nov, 2023', 'Oct, 2023', 'Dec, 2023', 'Oct, 2023', 'Jan, 2024', 'Oct, 2023', 'Feb, 2024', 'May, 2024', 'Oct, 2023', 'Jan, 2024', 'Sep, 2023', 'Jan, 2024', 'Jun, 2024', 'Oct, 2023', 'Jan, 2024', 'Oct, 2023', 'Feb, 2024', 'May, 2024', 'Oct, 2023', 'Jan, 2024', 'Sep, 2023', 'Jan, 2024', 'Jun, 2024', 'Oct, 2023', 'Jan, 2024', 'Oct, 2023', 'Feb, 2024', 'May, 2024', 'Oct, 2023', 'Jan, 2024', 'Sep, 2023', 'Jan, 2024', 'Jun, 2024', 'Oct, 2023', 'Jan, 2024', 'Oct, 2023', 'Feb, 2024', 'May, 2024', 'Oct, 2023', 'Jan, 2024', 'Sep, 2023', 'Jan, 2024', 'Jun, 2024', 'Oct, 2023', 'Jan, 2024', 'Oct, 2023', 'Feb, 2024', 'May, 2024', 'Oct, 2023', 'Jan, 2024', 'Sep, 2023', 'Jan, 2024', 'Jun, 2024', 'Oct, 2023', 'Jan, 2024', 'Oct, 2023', 'Feb, 2024', 'May, 2024', 'Oct, 2023', 'Jan, 2024', 'Sep, 2023', 'Jan, 2024', 'Jun, 2024', 'Oct, 2023', 'Jan, 2024', 'Oct, 2023', 'Feb, 2024', 'May, 2024', 'Oct, 2023', 'Jan, 2024'

In [5]:
len(converted_dates) # now i got the exact and equal lenght for all the columns

100

In [6]:
# Extract dates by selecting every second element (index 1, 3, 5, ...)
Dates = data[1::2]

print(Dates)
print("Total elements in Dates list:", len(Dates))


['11 months ago', 'Oct, 2023', 'Oct, 2023', '8 months ago', 'Oct, 2023', 'Jan, 2024', 'Jan, 2024', 'Nov, 2023', 'Oct, 2023', 'Dec, 2023', 'Oct, 2023', 'Jan, 2024', 'Oct, 2023', 'Feb, 2024', '9 months ago', 'Oct, 2023', 'Jan, 2024', 'Sep, 2023', 'Jan, 2024', '8 months ago', 'Oct, 2023', 'Jan, 2024', 'Oct, 2023', 'Feb, 2024', '9 months ago', 'Oct, 2023', 'Jan, 2024', 'Sep, 2023', 'Jan, 2024', '8 months ago', 'Oct, 2023', 'Jan, 2024', 'Oct, 2023', 'Feb, 2024', '9 months ago', 'Oct, 2023', 'Jan, 2024', 'Sep, 2023', 'Jan, 2024', '8 months ago', 'Oct, 2023', 'Jan, 2024', 'Oct, 2023', 'Feb, 2024', '9 months ago', 'Oct, 2023', 'Jan, 2024', 'Sep, 2023', 'Jan, 2024', '8 months ago', 'Oct, 2023', 'Jan, 2024', 'Oct, 2023', 'Feb, 2024', '9 months ago', 'Oct, 2023', 'Jan, 2024', 'Sep, 2023', 'Jan, 2024', '8 months ago', 'Oct, 2023', 'Jan, 2024', 'Oct, 2023', 'Feb, 2024', '9 months ago', 'Oct, 2023', 'Jan, 2024', 'Sep, 2023', 'Jan, 2024', '8 months ago', 'Oct, 2023', 'Jan, 2024', 'Oct, 2023', 'Feb, 2

In [7]:
# creating DataFrame
import pandas as pd
df = pd.DataFrame({"date" : Dates , "name" : Names , "city" : Cities , "review" : Reviews})
df

date              name                              city  \
0   11 months ago  Mousam  Guha Roy       Certified Buyer, Matialihat   
1       Oct, 2023            Ajin V         Certified Buyer, Balaghat   
2       Oct, 2023    bijaya mohanty        Certified Buyer, Baleshwar   
3    8 months ago   Prithivi Boruah          Certified Buyer, Bokajan   
4       Oct, 2023      Nikhil Kumar  Certified Buyer, Meerut Division   
..            ...               ...                               ...   
95      Oct, 2023     Urgain Norzin        Certified Buyer, New Delhi   
96      Jan, 2024       Surabhi Das           Certified Buyer, Raipur   
97      Sep, 2023     Nilesh  Yadav       Certified Buyer, Aurangabad   
98      Jan, 2024       SYAM SUNDER         Certified Buyer, Sangaria   
99   8 months ago    Abhishek Kumar   Certified Buyer, Siwan District   

                                               review  
0                                  Very niceREAD MORE  
1                       High quality camera😍READ MORE  
2   Just go for it.Amazing one.Beautiful camera wi...  
3       Camera Quality Is Improved Loving ItREAD MORE  
4   Switch from OnePlus to iPhone I am stunned wit...  
..                                                ...  
95  It's amazing nd also the delivery boy was very...  
96                                Awesome...READ MORE  
97  Best in class..! Superb camera and performance...  
98                         Camera is so goodREAD MORE  
99  awesome camera looks like Captured DSLR qualit...  

[100 rows x 4 columns]

In [8]:
#converting elements into lower string
df["review"] = df["review"].str.lower()
df

date              name                              city  \
0   11 months ago  Mousam  Guha Roy       Certified Buyer, Matialihat   
1       Oct, 2023            Ajin V         Certified Buyer, Balaghat   
2       Oct, 2023    bijaya mohanty        Certified Buyer, Baleshwar   
3    8 months ago   Prithivi Boruah          Certified Buyer, Bokajan   
4       Oct, 2023      Nikhil Kumar  Certified Buyer, Meerut Division   
..            ...               ...                               ...   
95      Oct, 2023     Urgain Norzin        Certified Buyer, New Delhi   
96      Jan, 2024       Surabhi Das           Certified Buyer, Raipur   
97      Sep, 2023     Nilesh  Yadav       Certified Buyer, Aurangabad   
98      Jan, 2024       SYAM SUNDER         Certified Buyer, Sangaria   
99   8 months ago    Abhishek Kumar   Certified Buyer, Siwan District   

                                               review  
0                                  very niceread more  
1                       high quality camera😍read more  
2   just go for it.amazing one.beautiful camera wi...  
3       camera quality is improved loving itread more  
4   switch from oneplus to iphone i am stunned wit...  
..                                                ...  
95  it's amazing nd also the delivery boy was very...  
96                                awesome...read more  
97  best in class..! superb camera and performance...  
98                         camera is so goodread more  
99  awesome camera looks like captured dslr qualit...  

[100 rows x 4 columns]

In [9]:
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import sent_tokenize
from nltk.tokenize import word_tokenize
from textblob import TextBlob
import string

nltk.download('stopwords')
nltk.download('punkt')
nltk.download('wordnet')

df["reviews"] = df["review"].astype(str).apply(lambda x: sent_tokenize(x))
df

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\suhai\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\suhai\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\suhai\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


date              name                              city  \
0   11 months ago  Mousam  Guha Roy       Certified Buyer, Matialihat   
1       Oct, 2023            Ajin V         Certified Buyer, Balaghat   
2       Oct, 2023    bijaya mohanty        Certified Buyer, Baleshwar   
3    8 months ago   Prithivi Boruah          Certified Buyer, Bokajan   
4       Oct, 2023      Nikhil Kumar  Certified Buyer, Meerut Division   
..            ...               ...                               ...   
95      Oct, 2023     Urgain Norzin        Certified Buyer, New Delhi   
96      Jan, 2024       Surabhi Das           Certified Buyer, Raipur   
97      Sep, 2023     Nilesh  Yadav       Certified Buyer, Aurangabad   
98      Jan, 2024       SYAM SUNDER         Certified Buyer, Sangaria   
99   8 months ago    Abhishek Kumar   Certified Buyer, Siwan District   

                                               review  \
0                                  very niceread more   
1                       high quality camera😍read more   
2   just go for it.amazing one.beautiful camera wi...   
3       camera quality is improved loving itread more   
4   switch from oneplus to iphone i am stunned wit...   
..                                                ...   
95  it's amazing nd also the delivery boy was very...   
96                                awesome...read more   
97  best in class..! superb camera and performance...   
98                         camera is so goodread more   
99  awesome camera looks like captured dslr qualit...   

                                              reviews  
0                                [very niceread more]  
1                     [high quality camera😍read more]  
2   [just go for it.amazing one.beautiful camera w...  
3     [camera quality is improved loving itread more]  
4   [switch from oneplus to iphone i am stunned wi...  
..                                                ...  
95  [it's amazing nd also the delivery boy was ver...  
96                              [awesome...read more]  
97  [best in class..!, superb camera and performan...  
98                       [camera is so goodread more]  
99  [awesome camera looks like captured dslr quali...  

[100 rows x 5 columns]

In [10]:
# Function to calculate polarity
def get_polarity(text):
    return TextBlob(text).sentiment.polarity

# Create a DataFrame with reviews
df = pd.DataFrame({'Review': Reviews})

# Apply polarity function
df['Polarity'] = df['Review'].apply(get_polarity)

# Display results
print(df[['Review', 'Polarity']])


                                               Review  Polarity
0                                  Very niceREAD MORE  0.350000
1                       High quality camera😍READ MORE  0.330000
2   Just go for it.Amazing one.Beautiful camera wi...  0.344444
3       Camera Quality Is Improved Loving ItREAD MORE  0.550000
4   Switch from OnePlus to iPhone I am stunned wit...  0.750000
..                                                ...       ...
95  It's amazing nd also the delivery boy was very...  0.433333
96                                Awesome...READ MORE  0.500000
97  Best in class..! Superb camera and performance...  0.729167
98                         Camera is so goodREAD MORE  0.500000
99  awesome camera looks like Captured DSLR qualit...  0.487500

[100 rows x 2 columns]


In [11]:
df["Polarity"].mean()

0.5080119444444444

 The value 0.508 suggests that, on average, the reviews are slightly positive since polarity values range from -1 (negative) to +1 (positive).